File: 2019/rust/day01/src/main.rs
use aoc::utils::parse_fields;
use std::io::{self, Read};

type Result<T> = ::std::result::Result<T, Box<dyn ::std::error::Error>>;

fn simple_fuel(mass: i64) -> i64 {
    std::cmp::max(0, mass / 3 - 2)
}

fn fuel(mass: i64) -> i64 {
    let fuel_needed = |&m: &i64| Some(simple_fuel(m)).filter(|&m| m > 0);

    std::iter::successors(Some(simple_fuel(mass)), fuel_needed).sum()
}

fn part1(masses: &Vec<i64>) -> i64 {
    masses.iter().cloned().map(simple_fuel).sum()
}

fn part2(masses: &Vec<i64>) -> i64 {
    masses.iter().cloned().map(fuel).sum()
}

fn main() -> Result<()> {
    let mut input = String::new();
    io::stdin().read_to_string(&mut input)?;
    let masses = parse_fields(input.trim(), '\n')?;

    println!("Part 1: {}", part1(&masses));
    println!("Part 2: {}", part2(&masses));
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn simple_fuel_tests() {
        assert!(simple_fuel(12) == 2);
        assert!(simple_fuel(14) == 2);
        assert!(simple_fuel(1969) == 654);
        assert!(simple_fuel(100756) == 33583);
    }

    #[test]
    fn fuel_tests() {
        assert!(fuel(14) == 2);
        assert!(fuel(1969) == 966);
        assert!(fuel(100756) == 50346);
    }
}

File: 2019/golang/day01/main.go
package main

import (
	"fmt"
	"julianandrews/adventofcode/aoc"
	"os"
)

func simple_fuel(mass int) int {
	val := mass/3 - 2
	if val > 0 {
		return val
	} else {
		return 0
	}
}

func fuel(mass int) int {
	total_mass := 0
	for mass > 0 {
		mass = simple_fuel(mass)
		total_mass += mass
	}
	return total_mass
}

func part_1(masses []int) int {
	total := 0
	for _, mass := range masses {
		total += simple_fuel(mass)
	}
	return total
}

func part_2(masses []int) int {
	total := 0
	for _, mass := range masses {
		total += fuel(mass)
	}
	return total
}

func main() {
	masses, err := aoc.GetInts()
	if err != nil {
		fmt.Fprintln(os.Stderr, err)
		os.Exit(1)
	}

	fmt.Println("Part 1: ", part_1(masses))
	fmt.Println("Part 2: ", part_2(masses))
}
